src/TokenGenerator.php

```diff
@@ -0,0 +1,145 @@
+			->setNotBefore(time() - 1)
+			->setExpiration(time() + 14*24*60*60)
+			->set("azp", $clientId)
+			->set("sub", $subject)
+			->set("jti", $this->generateJti())
+			->set("nonce", $nonce)
+			->set("at_hash", $tokenHash) //FIXME: at_hash should only be added if the response_type is a token
+			->set("c_hash", $tokenHash) // FIXME: c_hash should only be added if the response_type is a code
+			->set("cnf", array(
+				"jwk" => $jwks['keys'][0]
+			))
+			->withHeader('kid', $jwks['keys'][0]['kid'])
+			->sign($signer, $keychain->getPrivateKey($privateKey))
+			->getToken();
+		$result = $token->__toString();
+		return $result;
+	}
+	
+	public function respondToRegistration($registration, $privateKey) {
+		/*
+			Expects in $registration:
+			client_id
+			client_id_issued_at
+			redirect_uris
+			registration_client_uri
+		*/
+		$registration_access_token = $this->generateRegistrationAccessToken($registration['client_id'], $privateKey);
+
+		$registrationBase = array(
+			'response_types' => array("id_token token"),
+			'grant_types' => array("implicit"),
+			'application_type' => 'web',
+			'id_token_signed_response_alg' => "RS256",
+			'token_endpoint_auth_method' => 'client_secret_basic',
+			'registration_access_token' => $registration_access_token,
+		);
+		
+		$registration = array_merge($registrationBase, $registration);
+		return $registration;
+	}
+	
+	public function addIdTokenToResponse($response, $clientId, $subject, $nonce, $privateKey) {
+			if ($response->hasHeader("Location")) {
+			$value = $response->getHeaderLine("Location");
+			if (preg_match("/#access_token=(.*?)&/", $value, $matches)) {
+				$idToken = $this->generateIdToken(
+					$matches[1],
+					$clientId,
+					$subject,
+					$nonce,
+					$privateKey
+				);
+				$value = preg_replace("/#access_token=(.*?)&/", "#access_token=\$1&id_token=$idToken&", $value);				
+				$response = $response->withHeader("Location", $value);
+			} else if (preg_match("/code=(.*?)&/", $value, $matches)) {
+				$idToken = $this->generateIdToken(
+					$matches[1],
+					$clientId,
+					$subject,
+					$nonce,
+					$privateKey
+				);
+				$value = preg_replace("/code=(.*?)&/", "code=\$1&id_token=$idToken&", $value);
+				$response = $response->withHeader("Location", value);				
+			}
+		}
+		return $response;
+	}
+	///////////////////////////// HELPER FUNCTIONS \\\\\\\\\\\\\\\\\\\\\\\\\\\\\
+
+	private function generateJti() {
+		return substr(md5((string)time()), 12); // FIXME: generate unique jti values
+	}
+	
+	private function generateTokenHash($accessToken) {
+		$atHash = hash('sha256', $accessToken);
+		$atHash = substr($atHash, 0, 32);
+		$atHash = hex2bin($atHash);
+		$atHash = base64_encode($atHash);
+		$atHash = rtrim($atHash, '=');
+		$atHash = str_replace('/', '_', $atHash);
+		$atHash = str_replace('+', '-', $atHash);
+
+		return $atHash;
+	}
+
+	private function getJwks() {
+        $key = $this->config->getKeys()->getPublicKey();
+        $jwks = new Jwks($key);
```

Comment:
Shouldn't the created `$jwks` be returend here?

Reply:
maybe - I did mapped it back to a simple array because that is what I expect/need in the application layer.

Reply:
Oh, no wait. I misread... Somehow the indent made me think that nothing was returned and the last lines was inside a callback. Nvrmnd... 😵